### ISSUE TITLE
[Bug]: Incorrect cropping range in crop_and_resize when input boxes has non-parallel opposite sides

### ISSUE THREAD

**Issue by Mine268** [kornia/kornia#3560]
### 🐛 Describe the bug

`kornia.geometry.transform.crop_and_resize` behaves abnormally when oppsite sides of argument `boxes` are unparallel. This would result the cropped area cannot match the exact area of `boxes` in the `input_tensor`.

---

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/85843dc7-3032-4d4d-85e2-f95000423d09" />
<br/>Please download this image as checkboard.png for reproducing the bug.

---

<img width="256" height="256" alt="Image annotated with cropping area" src="https://github.com/user-attachments/assets/28c578bd-022e-4632-867f-011993d99444" />
<br/>Image annotated with cropping 

---

<img width="128" height="128" alt="The cropping result" src="https://github.com/user-attachments/assets/eebc80b2-ec47-487e-9455-b0d4f8d6fc51" />
<br/>The incorrect cropping result via kornia's crop_and_resize

---

<img width="128" height="128" alt="Image" src="https://github.com/user-attachments/assets/86cca40d-f0dc-4944-96e8-faabbf56c8e4" />
<br/>The correct cropping result by opencv

### 🔄 Steps to Reproduce

```bash
1. read the checkboard image
2. designate 4 corners of cropping bbox by `corner_idx`, the left-top corner of corresponding block is selected, order matches the signature of `crop_and_resizez
3. invoke `kornia.geometry.transform.crop_and_resize` to crop the image
4. get the result as description

For detail procedure, please refer to Minimal Code Example.
```

### 💻 Minimal Code Example

```python
Minimal python code to reproduce the bug.

Please save the origin picture depicted in description.


import cv2
import numpy as np
import torch
import kornia

# Read the checkboard, 8*8, 32px*32px for each block
img = cv2.imread("./checkerboard.png")

# Four corners for perspective cropping,
# for corners of the area is the left-up corner of corresponding block
corner_idx = [29, 23, 56, 45]
corner_coord = []

for cx in corner_idx:
    j = (cx - 1) // 8
    i = (cx - 1) % 8
    corner_coord.append((i * 32, j * 32))

# draw the cropping box
for cx in range(len(corner_coord)):
    p1 = corner_coord[cx]
    p2 = corner_coord[(cx + 1) % len(corner_coord)]
    cv2.line(
        img,
        p1,
        p2,
        (0, 0, 255),
        2,
    )
    cv2.putText(
        img,
        str(cx),
        (p1[0], p1[1] - 5),
        cv2.FONT_HERSHEY_SIMPLEX,
        0.5,
        (0, 0, 255),
        1,
    )

# Show the cropping area in origin image
cv2.imshow("cropping perspective area", img)
cv2.imwrite("cropping_area.png", img)

# Reproduce the bug
img = torch.from_numpy(img)
img = img.permute(2, 0, 1).unsqueeze(0) / 255 # [1,3,256,256]

img_result = kornia.geometry.transform.crop_and_resize(
    img,
    torch.tensor(corner_coord).unsqueeze(0), # [1,4,2]
    [128, 128]
)
img_result = (img_result[0].permute(1, 2, 0).numpy() * 255).astype(np.uint8)

cv2.imshow("cropped", img_result)
cv2.imwrite("cropping_result.png", img_result)

# Desired result
persp_mat = cv2.getPerspectiveTransform(
    np.array(corner_coord).astype(np.float32),
    np.array([
        [0, 0], [128, 0], [128, 128], [0, 128]
    ]).astype(np.float32)
)
img_result2 = cv2.warpPerspective(
    img[0].permute(1, 2, 0).numpy(),
    persp_mat,
    [128, 128]
)
img_result2 = (img_result2 * 255).astype(np.uint8)
cv2.imshow("cropped opencv", img_result2)
cv2.imwrite("cropping_result2.png", img_result2)

cv2.waitKey(0)
```

### ✅ Expected behavior

The area within the red border area should be cropped and rectified into the square image under perspective transformation:

<img width="128" height="128" alt="Image" src="https://github.com/user-attachments/assets/9710e801-07cf-4684-a972-269872f0a954" />

### ❌ Actual behavior

The wrong area is cropped and transformed:

<img width="128" height="128" alt="Image" src="https://github.com/user-attachments/assets/b79f78af-1d05-42a3-be06-0ca7730f6993" />

### 🔧 Environment

```shell
- Kornia version: 0.8.2
- PyTorch version: 2.10.0
- Python version: 3.12.12
- OS (e.g., Linux, macOS, Windows): found on ubuntu 20.04 with cuda, verified on macOS
- Installation method (pip, conda, from source): pip/uv
- CUDA/cuDNN version (if applicable): no
- GPU model (if applicable): no
(above behaviour is verified on macOS with cpu, but I assume that the cuda will not affect the result)

script output:

Collecting environment information...
PyTorch version: 2.9.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.2 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.0.13.5)
CMake version: version 3.30.2
Libc version: N/A

Python version: 3.12.12 | packaged by conda-forge | (main, Oct 22 2025, 23:34:53) [Clang 19.1.7 ] (64-bit runtime)
Python platform: macOS-26.2-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M2

Versions of relevant libraries:
[pip3] Could not collect
[conda] numpy                     2.3.2                    pypi_0    pypi
```

### 📝 Additional context

This issue is found during using kornia with my machine learning project.

### 🤝 Contribution Intent

- [x] I plan to submit a PR to fix this bug
- [x] I'm reporting this bug but not planning to fix it

**Comment by navin-oss** [comment]
Hi @Mine268, I’d like to investigate and work on this issue.

Plan:
* Reproduce the behavior locally using the provided example
* Compare Kornia’s `crop_and_resize` math with OpenCV’s perspective transform
* Debug the homography/corner mapping logic for non-parallel quadrilaterals
* Propose a fix with tests to ensure correct cropping for general 4-point boxes
Please assign this issue to me if available.

**Comment by Mine268** [comment]
Thanks @navin-oss for your kindness, I'd like to assign this to you.

I have investigated the kornia code for the reason behind this, and I think I might find the cause.

Navigate to `kornia/geometry/transform/crop2d.py, line 307`, where the `patches` is sampled by `warp_affine` function:
```python
patches: torch.Tensor = warp_affine(
    input_tensor,
    dst_trans_src[:, :2, :],
    out_size,
    mode=mode,
    padding_mode=padding_mode,
    align_corners=align_corners,
)
```

`dst_trans_src` is the perspective matrix computed by `get_perspective_transform (line 262)`, where the four corners are taken as arguments. In the case that "non-parallel quadrilaterals", last row of this perspective matrix is not [0,0,1], the `dst_trans_src` is not an affine matrix, thus incompatible with `warp_affine`.

I suspect that whether `warp_perspective` would fix this issue.

PS: I am not familar with the workflows (PR, issue, ...), and i hope above info can be helpfull for you and kornia.
PPS: I was trying to implement an efficient cropping procedure under perspective normalization on image, and found this issue. Please feel free to contact me for more details.

**Comment by Rajat-Arora** [comment]
Hi @sidd-27, I went through the comments and would like to take this up. Could you assign this to me?

**Comment by aarush-github** [comment]
Hey @Mine268 @sidd-27, Thanks for explaining the issue is such great detail.
I would love to take this up!

Proposed Fix:
The current implementation uses 'warp_affine' which discards the perspective components of the transformation matrix by slicing it to 2x3. This leads to incorrect results for non-parallel boxes.

1. Update crop_and_resize to use warp_perspective instead of warp_affine.

2. Pass 3x3 transformation matrix to ensure perspective distortions are correctly handled.

3. Add a tests with non-parallel quadrilateral boxes to verify the fix.

Could you please assign this to me?

**Comment by Mine268** [comment]
> Hi [@Mine268](https://github.com/Mine268), I’d like to investigate and work on this issue.
> 
> Plan:
> 
> * Reproduce the behavior locally using the provided example
> * Compare Kornia’s `crop_and_resize` math with OpenCV’s perspective transform
> * Debug the homography/corner mapping logic for non-parallel quadrilaterals
> * Propose a fix with tests to ensure correct cropping for general 4-point boxes
>   Please assign this issue to me if available.

@navin-oss hi, have you already made a pull request for the fixing?

**Comment by navin-oss** [comment]
I need to check this

**Comment by shijianjian** [comment]
Hi @Rajat-Arora and @aarush-github, are you working on this?

**Comment by Mine268** [comment]
> Hi [@Rajat-Arora](https://github.com/Rajat-Arora) and [@aarush-github](https://github.com/aarush-github), are you working on this?

i suppose they don't... it has been a while orz. i'd fix it myself if graduation weren't keeping me so busy.

**Comment by Meggison** [comment]
@FloodExLLC need to check with other contributors related to the issue before linking a PR to it
